ACE/ace/SOCK_Dgram.cpp

```diff
@@ -672,7 +672,63 @@ ACE_SOCK_Dgram::make_multicast_ifaddr (ip_mreq *ret_mreq,
       // This port number is not necessary, just convenient
       ACE_INET_Addr interface_addr;
       if (interface_addr.set (mcast_addr.get_port_number (), net_if) == -1)
-        return -1;
+        {
+#ifdef ACE_WIN32
+          IP_ADAPTER_ADDRESSES tmp_addrs;
+          // Initial call to determine actual memory size needed
+          DWORD dwRetVal;
+          ULONG bufLen = 0;
+          if ((dwRetVal = ::GetAdaptersAddresses (AF_INET,
+                                                  0,
+                                                  0,
+                                                  &tmp_addrs,
+                                                  &bufLen)) != ERROR_BUFFER_OVERFLOW)
+            return -1; // With output bufferlength 0 this can't be right.
+
+          // Get required output buffer and retrieve info for real.
+          PIP_ADAPTER_ADDRESSES pAddrs;
+          char *buf;
+          ACE_NEW_RETURN (buf,
+                          char[bufLen],
+                          -1);
+          pAddrs = reinterpret_cast<PIP_ADAPTER_ADDRESSES> (buf);
+          if ((dwRetVal = ::GetAdaptersAddresses (AF_INET,
+                                                  0,
+                                                  0,
+                                                  pAddrs,
+                                                  &bufLen)) != NO_ERROR)
+            {
+              delete[] buf; // clean up
+              return -1;
+            }
+
+          interface_addr = ACE_INET_Addr(); // initialize
+          int set_result = -1;
+          while (pAddrs && set_result == -1)
+            {
+              if (ACE_OS::strcmp(ACE_TEXT_ALWAYS_CHAR(net_if), pAddrs->AdapterName) == 0 ||
+                  ACE_OS::strcmp(ACE_TEXT_ALWAYS_CHAR(net_if), ACE_Wide_To_Ascii(pAddrs->FriendlyName).char_rep()) == 0)
```

Comment:
Why compare two wchar back to char to compare, why not compare as wchar?

Reply:
I copied & modified most of this chunk of code from make_multicast_ifaddr6, but I agree, it could be better. If the values being copied had been the same, I would have made a separate function to avoid duplication. I'll fix these issues in both locations.